install/kubernetes/sidecar-template.yaml

```diff
@@ -0,0 +1,90 @@
+initContainers:
+- name: istio-init
+  image: "gcr.io/istio-testing/proxy_init:3101ea9d82a5f83b699c2d3245b371a19fa6bef4"  
+  args: ["-p","15001", "-u", "1337"]
+  imagePullPolicy: "IfNotPresent"
+  securityContext:
+    capabilities:
+      add:
+      - NET_ADMIN
+    priviledged: true
+  restartPolicy: Always
+containers:
+- name: istio-proxy
+  image: gcr.io/istio-testing/proxy_debug:3101ea9d82a5f83b699c2d3245b371a19fa6bef4
+  args: 
+  - proxy
+  - sidecar
+  - -v
+  - "2"
+  - --configPath
+  - /etc/istio/proxy
+  - --binaryPath
+  - /usr/local/bin/envoy
+  - --serviceCluster
+  {{ if eq .ServiceCluster "" -}}
+  - {{ "istio-proxy" }}
+  {{ else -}}
+  - {{ printf "%s" .ServiceCluster }}
+  {{ end -}}
+  - helloworld
+  - --drainDuration
+  - 45s
+  - --parentShutdownDuration
+  - 1m0s
```

Comment:
these all should be customizable.

Reply:
This is consumed directly by the initializer/webhook. Golang template is used to handle the few corner cases where injected resources depends on the content of the injectee, otherwise it would be a fixed list of resources a. la https://github.com/kelseyhightower/kubernetes-initializer-tutorial/blob/master/configmaps/envoy-initializer.yaml.

Further user customization and tweaking is better served by another layer on top of this file. The current plan is to use the existing `istioctl kube-inject` flags to generate `sidecar-template.yaml`. This decouples the new data-driven configuration from the older parameterized model while preserving some level of backwards compat. 

Things like jsonnet or custom CI/CD solutions would likely plug into this outer. The only requirement is that that adhere to the basic format of list of containers, initContainers, and volumes.

Reply:
That makes sense. You want "curry" the configmap parameter application from the the pod/deployment parameters. I think it would be cleaner to use same templating mechanism for both steps. Hence, my suggestion to use language-independent scripting like jsonnet to do both 1) generate sidecar-template.jsonnet and 2) apply sidecar-template.jsonnet to pods. Then we support three options for CI/CD:
1. Commit k8s objects. Run kube-inject & injector dynamically.
2. Commit k8s objects and sidecar-template. Run injector dynamically.
3. Commit modified k8s objects. Run nothing.
All three options are plausible.